File: solutions/java/queue/step05_size/Queue.java
import java.util.ArrayList;
import java.util.List;

public class Queue<T> {
    private final List<T> items;
    private int head;

    public Queue() {
        this.items = new ArrayList<>();
        this.head = 0;
    }

    public void enqueue(T value) {
        items.add(value);
    }

    public T dequeue() {
        if (isEmpty()) {
            return null;
        }
        T value = items.get(head);
        head += 1;
        if (head > items.size() / 2) {
            items.subList(0, head).clear();
            head = 0;
        }
        return value;
    }

    public T front() {
        if (isEmpty()) {
            return null;
        }
        return items.get(head);
    }

    public int size() {
        return items.size() - head;
    }

    public boolean isEmpty() {
        return size() == 0;
    }
}

File: solutions/python/queue/05_size/queue.py
class Queue:
    def __init__(self) -> None:
        self._items = []
        self._head = 0

    def enqueue(self, value) -> None:
        self._items.append(value)

    def dequeue(self):
        if self.is_empty():
            return None
        value = self._items[self._head]
        self._head += 1
        if self._head > len(self._items) // 2:
            self._items = self._items[self._head :]
            self._head = 0
        return value

    def front(self):
        if self.is_empty():
            return None
        return self._items[self._head]

    def size(self) -> int:
        return len(self._items) - self._head

    def is_empty(self) -> bool:
        return self.size() == 0

